src/main/java/LeaveGroup/LeaveGroupDsGateway.java

```diff
@@ -0,0 +1,30 @@
+package LeaveGroup;
+
+import Entities.Group;
+import Entities.User;
+
+public interface LeaveGroupDsGateway {
+    boolean groupExists(String groupname);
+    User getUser(String username);
+    Group getGroup(String groupname);
+
+    /**
+     * Updates the current user after user's groups list has changed. Saves to repository.
+     * @param username the username of the current user
+     */
+    void updateUser(String username);
+
+    /**
+     * Updates the group after group's members list has changed. Saves to repository.
+     * @param groupname the name of the group
+     */
+    void updateGroup(String groupname);
+    boolean userInGroup(String username, String groupName);
+
+    /**
+     * @param username the username of the current user
+     * @param groupName the name of the group
+     * @return whether the user's members list contains the group's name.
+     */
+    boolean groupInUser(String username, String groupName);
+}
```

Comment:
This will be fixed when we make one general gateway, but for you own use case have you considered combining your two gateways? they share similar functions such as getUser, groupExists, and so on. 

Also, for updateGroup(String groupname) how is the repo going to know how to update the group (i.e. how will it know that the users have changed)? especially if it just has group name. Maybe I am missing something.

Reply:
My two dsGateways are the exact same, and I think they're pretty identical to Sharon and Ipek's. We talked to Paul about this issue -- he didn't really answer the question about HOW we could combine gateways across usecases, but he said that it wasn't an issue for different use cases to have similar methods in their gateways since each use case should be totally independent from the rest. He said that as long as the repo implements all the gateways, there is no problem.

I'm not totally convinced by this answer, and I think I'll follow up with him again.

Reply:
> Also, for updateGroup(String groupname) how is the repo going to know how to update the group (i.e. how will it know that the users have changed)? especially if it just has group name. Maybe I am missing something.

Oh you're so right actually thank you for catching this! I forgot to update this method after we figured out the current user thing.. the method should have the current user username and the groupname as parameters.

The repository has the ability to find the group object from its groupname (because it contains all Groups). So the updateGroup() method implemented by the repo will be able to get the Group object from the groupname and then mutate its members list and then call the save() method in the repo to record this change. Does this make sense, feel free to correct me if I'm not understanding the repo correctly.

Reply:
> My two dsGateways are the exact same, and I think they're pretty identical to Sharon and Ipek's. We talked to Paul about this issue -- he didn't really answer the question about HOW we could combine gateways across usecases, but he said that it wasn't an issue for different use cases to have similar methods in their gateways since each use case should be totally independent from the rest. He said that as long as the repo implements all the gateways, there is no problem.
> 
> I'm not totally convinced by this answer, and I think I'll follow up with him again.

That's actually so interesting hmmm. I guess it would make sense to have each use case have its own dsGateway otherwise our application would be way too dependent on our General DsGateway. If we changed something it might cause errors in all use cases rather than just one. Okay that makes sense thank you for the clarification!

Reply:
> 

Great that makes sense! Good implementation idea! Having the username as the input would make it easier to remove it from the group's member list, or application list. It also seems like the best way to do so while also maintaining clear architecture and not passing around entities.